moveit_reach_plugins/src/evaluation/manipulability_moveit.cpp

```diff
@@ -141,9 +169,93 @@ double ManipulabilityRatio::calculateScore(const Eigen::MatrixXd& jacobian_singu
   return jacobian_singular_values.minCoeff() / jacobian_singular_values.maxCoeff();
 }
 
+bool ManipulabilityNormalized::initialize(XmlRpc::XmlRpcValue& config)
+{
+  bool ret = ManipulabilityMoveIt::initialize(config);
+
+  if (config.hasMember("exclusion_list") && config["exclusion_list"].getType() == XmlRpc::XmlRpcValue::TypeArray)
+  {
+    for (std::size_t i = 0; i < config["exclusion_list"].size(); ++i)
+    {
+      std::string excluded_link = static_cast<std::string>(config["exclusion_list"][i]);
+      exclusion_list_.push_back(excluded_link);
+    }
+
+    if (exclusion_list_.empty())
+    {
+      ROS_ERROR_STREAM("Exclusion list is empty");
+      return false;
+    }
+  }
+
+  calculateCharacteristicLength();
+  return ret;
+}
+
+double ManipulabilityNormalized::calculateScore(const Eigen::MatrixXd& jacobian_singular_values)
+{
+  if (characteristic_length_ == 0)
+    throw std::runtime_error("The model must have a nonzero characteristic length");
+
+  return ManipulabilityMoveIt::calculateScore(jacobian_singular_values) / characteristic_length_;
+}
+
+/*
+1. Stay as is, but add inclusion or exclusion lists
+  - boom_crane exclude joint_link_1 and joint_link_3
+2. Add random sampling
+3. Reverse IK? "On the Optimum Dimensioning of Robotic Manipulators"
+*/
+
+double ManipulabilityNormalized::calculateCharacteristicLength()
```

Comment:
This is probably a useful function to have as a utility. I would suggest moving this function out of this particular class into a stand-alone helper function (defined in `manipulability_moveit.h`) and change the signature to:

```c++
double calculateCharacteristicLength(moveit::core::RobotModelConstPtr model, const std::string& motion_group, const std::vector<std::string>& exclusion_links = {})
```